2022/01/how-to-use-regex-groups-quantifier-in-postgresql.md

```diff
@@ -0,0 +1,86 @@
+---
+author: "Selvakumar Arumugm"
+title: "How to use regex groups quantifier in PostgreSQL"
+date: 2022-01-07
+tags:
+- postgres
+- sql
+- regex
+- hl7
+github_issue_number: 
+---
+
+I recently encountered a situation where it was necessary to extract address content from a
+specific format (HL7 V2) of text on a PostgreSQL table column. This became possible through applying a regex as the address contains a standard format that regex can match. The following example will be used to demonstrate the process: 
+
+```
+||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|
+```
+### Perl Regex Pattern
+
+In order to manipulate our example, the address section needs to be extracted from the table. The sections and fields are delimited by pipes ( | ) and carets ( ^ ). Uniquely, the address contains eight carets. Therefore, a regex group with alphanumeric and caret repetition is required in order to match our address. The regex is tested through the grep command perl regex option (-P, --perl-regexp) and successfully extracts the address section from the content.
+
+```bash
+$ content='||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|'
+
+$ echo $content | grep -Po '([A-Za-z0-9 ]*\^){8}'
+123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^
+```
+
+### PostgreSQL Regex
+
+PostgreSQL's `regexp_matches` function supports the pattern extraction by matching data from the content. While the same perl regex pattern was added to the `regexp_matches` function, instead of all eight groups, only the eighth value was returned. 
+
+```postgres
+=# SELECT REGEXP_MATCHES('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|', E'([A-Za-z0-9 ]*\\^){8}[A-Za-z0-9 ]*', 'g');
+ regexp_matches 
+----------------
+ {^}
+(1 row)
+```
+
+More generally, our query is returning the Nth matching group instead of returning all matching groups until the Nth regex group. So if we try to fetch the text matching with 3 groups, the quantifier three will return the third field of all match sections instead of the third group itself.
```

Comment:
The reason this is different from your `grep` is that `grep -o` is like putting a `(...)` around the whole match and returning it, and there's nothing similar to that in Postgres or Perl regexes natively.

Reply:
Good to know